### DISCUSSION TITLE
Minimum time problems

### DISCUSSION THREAD

**Discussion by dynaTenCo** [infiniteopt/InfiniteOpt.jl#188] [Q&A] [answered]
Can anyone provide links or examples of solving a minimum time problem using InfiniteOpt.jl?
For example, the hovercraft example in the documentation - if the objective is to complete the path in the shortest amount of time (and we don't care about the thrust), how would we formulate the objective, constraints, etc? The final time and the way point times also become variables, so not sure how to formulate the problem.
Thanks in advance.

**Comment by pulsipher** [comment, marked as answer]
Hi there! Welcome to the InfiniteOpt discussion forum!

Minimum time problems can be reformulated into a standard optimal control problem. A nice video tutorial is given [here](https://apmonitor.com/do/index.php/Main/RocketOptimization) in the context of an optimal rocket launch problem where we want to minimize the final time for the rocket to arrive. 

The key reformulation trick is to scale the total time by the final time with `τ = t / tf` and let `tf` be a finite variable.

The script below implements the same rocket example in InfiniteOpt:
```julia
using InfiniteOpt, Ipopt, Plots

# Define the model
model = InfiniteModel(Ipopt.Optimizer)

# Setup up the normalized time parameter τ = t / tf
@infinite_parameter(model, τ in [0, 1], num_supports = 101)

# Create the variables
@variable(model, -1.1 <= u <= 1.1, Infinite(τ), start = 0)
@variable(model, s, Infinite(τ), start = 0)
@variable(model, 0 <= v <= 1.7, Infinite(τ), start = 0)
@variable(model, m >= 0.2, Infinite(τ), start = 1)
@variable(model, 0.1 <= tf <= 100, start = 1)

# Set the objective
@objective(model, Min, tf)

# Define the ODEs
@constraint(model, ∂(s, τ) == tf * v)
@constraint(model, m * ∂(v, τ) == tf * (u - 0.2 * v^2))
@constraint(model, ∂(m, τ) == tf * (-0.01 * u^2))

# Set the initial conditions
@constraint(model, s(0) == 0)
@constraint(model, v(0) == 0)
@constraint(model, m(0) == 1)

# Set terminal constraints
@constraint(model, s(1) == 10)
@constraint(model, v(1) == 0)

# Optimize and get the results
optimize!(model)
opt_u = value(u)
opt_s = value(s)
opt_m = value(m)
opt_v = value(v)
opt_tf = value(tf)

# Get the scaled times
ts = value(τ) * opt_tf

# Plot the optimal trajectories
plot(ts, [opt_s opt_v opt_m opt_u], layout = (4, 1), legend = false, 
     xlabel = "Time", ylabel = ["Position" "Velocity" "Mass" "Force"],
     xlims = (0, opt_tf))
```
![image](https://user-images.githubusercontent.com/33848311/145079148-c0793964-caab-49b5-8b05-00e8a04c921b.png)

**Reply by dynaTenCo** [reply]
Thank you for the answer. InfiniteOpt seems to be exactly what I was looking for. Look forward to working with it.

**Reply by pulsipher** [reply]
Glad to be of help, please feel free to reach out again with any further questions or comments.

**Reply by dynaTenCo** [reply]
My understanding is that the discretized optimization problem is solved using orthogonal collocation method. This should result in a piece-wise constant solution for optimal u(t). But the plot of force looks like a continuous function. Is this just an artifact of the plot function interpolating values (at the places where it ramps up or down)?

**Reply by pulsipher** [reply]
We can solve it via finite difference methods or orthogonal collocation (the default used above is backward finite difference). In this case, u(t) effectively becomes a piece-wise linear function  (the plot interpolates). However, in general, my understanding that the control function u(t) need not necessarily be a piece-wise function and can be a general continuous function as well (I can think of a number of examples where this is the case). This behavior is independent of our selected numerical method to approximate the state derivatives (e.g., orthogonal collocation).  These numerical methods simply provide a discrete approximation of the underlying continuous-time optimal solution.

**Reply by pulsipher** [reply]
I should also note that these are all solved simultaneously as is explained in "Nonlinear Programming: Concepts, Algorithms, and Applications to Chemical Processes"

**Reply by dynaTenCo** [reply]
OK, thanks for the info. I will study the documentation in more detail.
